analysis_notebooks/Attribute Analysis.ipynb

```diff
@@ -981,6 +984,296 @@
+    "## Setup\n",
+    "<a id=\"att_dis_setup\"></a>"
+   ],
+   "cell_type": "markdown",
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "hide_toggle(for_next=True, toggle_text='Show/hide att_dis code')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "categories = dataset.categories\n",
+    "names = dataset.labels_to_names\n",
+    "distances = pickle.load(open(\"results/{}/att_dis.pkl\".format(folder_name), \"rb\"))\n",
+    "topn = 10\n",
+    "\n",
+    "attr_dists = []\n",
+    "attr_distr_dists = []\n",
+    "attr_file_paths = []\n",
+    "attr_indices = []\n",
+    "p_values = []\n",
+    "ratio_idxs = []\n",
+    "xaxis = []\n",
+    "\n",
+    "for i in range(len(categories)):\n",
+    "    cat_dists = []\n",
+    "    cat_distr_dists = []\n",
+    "    cat_file_paths = []\n",
+    "    cat_indices = []\n",
+    "    cat_p_values = []\n",
+    "    largest_ratio = -1\n",
+    "    ratio_idx = -1\n",
+    "    for j in range(num_attrs):\n",
+    "        a_dist = np.array([ (chunk[0] / np.sqrt(chunk[1]*chunk[2])) for chunk in distances[i][j] if chunk[2] != 0 and chunk[1] != 0])\n",
+    "        a_file = np.array([ chunk[3] for chunk in distances[i][j] if chunk[2] != 0 and chunk[1] != 0])\n",
+    "        a_index = np.array([ chunk[4] for chunk in distances[i][j] if chunk[2] != 0 and chunk[1] != 0])\n",
+    "        b_dist = []\n",
+    "        b_file = []\n",
+    "        b_index = []\n",
+    "        for k in range(num_attrs):\n",
+    "            b_dist.extend([(chunk[0] / np.sqrt(chunk[1]*chunk[2])) for chunk in distances[i][k] if chunk[2] != 0 and chunk[1] != 0 and j != k])\n",
+    "            b_file.extend([chunk[3] for chunk in distances[i][k] if chunk[2] != 0 and chunk[1] != 0 and j != k])\n",
+    "            b_index.extend([chunk[4] for chunk in distances[i][k] if chunk[2] != 0 and chunk[1] != 0 and j != k])\n",
+    "        \n",
+    "        t, p = stats.ttest_ind(a_dist, b_dist, equal_var=False)\n",
+    "\n",
+    "        cat_p_values.append(p)\n",
+    "        cat_distr_dists.append([a_dist, b_dist])\n",
+    "        cat_dists.append([np.mean(a_dist), np.mean(b_dist)])\n",
+    "        cat_file_paths.append([a_file, b_file])\n",
+    "        cat_indices.append([a_index, b_index])\n",
+    "\n",
+    "        a_mean = np.mean(a_dist)\n",
+    "        b_mean = np.mean(b_dist)\n",
+    "        if not np.isnan(a_mean) and not np.isnan(b_mean) and not len(a_dist) < 5 and not len(b_dist) < 5:\n",
+    "            if (a_mean >= b_mean and a_mean/b_mean > largest_ratio and p < 0.05):\n",
+    "                largest_ratio = a_mean/b_mean\n",
+    "                ratio_idx = j\n",
+    "            if (a_mean < b_mean and b_mean/a_mean > largest_ratio and p < 0.05):\n",
+    "                largest_ratio = b_mean/a_mean\n",
```

Comment:
does this mean that for one object, only one ratio can be shown? what if group 1 is super different than the rest, but group 3 is also super different than the rest for the same object. both should be able to be shown for this object